functorch/_src/aot_autograd.py

```diff
@@ -1457,7 +1499,36 @@ def compiled_function(*args):
         else:
             return fw_outs
 
-    return compiled_function
+    if not config.debug_assert:
+        return compiled_function
+
+    flat_requires_grad = [a.requires_grad if isinstance(a, Tensor) else None for a in flat_args]
+
+    @wraps(compiled_function)
+    def debug_compiled_function(*args):
+        # TODO: Check aliasing relationships
+        # TODO: Check strides for metadata mutation
+        # (NB: ideally, this logic is factored out of this function and
+        # you move these debug checks there)
+
+        # Check requires grad.  Bad case is when we compiled with
+        # requires_grad = False, but input requires_grad = True
+        # (vice versa is OK; we compute a gradient and then throw
+        # it away when it hits the input.)
```

Comment:
This is definitely reasonable, but it feels like a compilation policy decision? Maybe we should include this in a section of compilation policy (when we add one / doesn't need to be in this PR).

There's the degenerate case where a user runs a function once with a bunch of requires_grad=True inputs, and then multiple times with many of the inputs set to requires_grad=False. Then we're compiling a backward function that does a bunch of compute for no reason. Although I can see this case not mattering much in practice, and compiling fewer times being more important.

Reply:
There are two parts: is it sound to use a requires_grad = True graph when requires_grad = False, and then the policy decision.

The soundness argument is unconditionally true. It's just like how we have derivative formulas that say how to calculate derivatives for all inputs, but you may not actually use all of them. The assertion here only cares about whether or not something is sound: it needs to assert when something is unsound (not when you've made a suboptimal policy decision.)

You've accuracy described the policy problem. But in fact, the "compile more, avoid dead compute" policy IS, hypothetically, what is currently implemented. This is because Dynamo installs guards on all tensors which require `requires_grad` to match exactly. However, in an earlier CI run, I tried the strong invariant, and it actually fails in practice. So there is some fishy going on with the guards, which may warrant some further exploration. But maybe for now we should just log a warning when this happens.